### PR TITLE
Fix some tests to allow ambiguous error cases.

### PR DESCRIPTION
**Pull request by jdashg**
There are a number of tests that assume the most obvious GL errors will be generated. However, a single function call can match multiple error generation criteria.

I add functions that take a list of possible errors, and make sure that the generated error in the given list.

### CODE REVIEW

**Comment by zhenyao** [comment]
This looks good to me.  I agree the conformance tests shouldn't over-testing the error cases.

File: sdk/tests/conformance/context/context-lost-restored.html [hunk 1]
```diff
@@ -248,7 +248,8 @@
     // Extension must still be lost.
     var tex = gl.createTexture();
     gl.bindTexture(gl.TEXTURE_2D, tex);
-    shouldGenerateGLError(gl, gl.INVALID_ENUM, "gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, 1, 1, 0, gl.RGBA, gl.FLOAT, null)");
+    wtu.shouldGenerateGLErrorIn(gl, [gl.INVALID_ENUM, gl.INVALID_OPERATION],
```
Comment (kenrussell) [inline sdk/tests/conformance/context/context-lost-restored.html]: This change looks incorrect. The only error this call is supposed to generate is INVALID_ENUM because of the use of the gl.FLOAT type, since the OES_texture_float extension is lost when the context is restored.

File: sdk/tests/conformance/extensions/ext-sRGB.html [hunk 2]
```diff
@@ -125,9 +125,7 @@
   var err = gl.getError();
   if (err == gl.NO_ERROR) {
     testFailed("should NOT be able to create type " + formatName);
-  } else if (err == gl.INVALID_OPERATION) {
-    testFailed("should return gl.INVALID_ENUM for type " + formatName);
-  } else if (err == gl.INVALID_ENUM) {
+  } else if (err == gl.INVALID_ENUM || err == gl.INVALID_OPERATION) {
```
Comment (kenrussell) [inline sdk/tests/conformance/extensions/ext-sRGB.html]: This change also looks incorrect. Per https://www.khronos.org/opengles/sdk/docs/man/xhtml/glTexImage2D.xml the only error condition we should be hitting is that for INVALID_ENUM, since the format isn't an accepted value.

File: sdk/tests/conformance/extensions/oes-texture-float.html [hunk 3]
```diff
@@ -157,7 +157,8 @@
     }
     gl.texImage2D(gl.TEXTURE_2D, 0, format, width, height, 0, format, gl.FLOAT, data);
     if (expectFailure) {
-        wtu.glErrorShouldBe(gl, gl.INVALID_ENUM, "floating-point texture allocation must be disallowed if OES_texture_float isn't enabled");
+        wtu.glErrorShouldBe(gl, [gl.INVALID_ENUM, gl.INVALID_OPERATION],
```
Comment (kenrussell) [inline sdk/tests/conformance/extensions/oes-texture-float.html]: As above, the only error that should be possible from the texImage2D call above is INVALID_ENUM, since gl.FLOAT isn't supported because the OES_texture_float extension hasn't been enabled.

File: sdk/tests/conformance/extensions/ext-sRGB.html (hunk 2 above)
Comment (kenrussell) [inline sdk/tests/conformance/extensions/ext-sRGB.html]: Looking more closely, it looks like INVALID_VALUE could be generated here since internalformat isn't an accepted format, but INVALID_OPERATION should not be able to be generated.

File: sdk/tests/conformance/extensions/oes-texture-half-float.html [hunk 4]
```diff
@@ -232,19 +232,20 @@
...
-          ". Expect " + (expectSuccess ? "Success" : "Failure"));   
+          ". Expect " + (expectSuccess ? "Success" : "Failure"));
 
     var texture = allocateTexture();
     var width = 2;
     var height = 2;
     gl.texImage2D(gl.TEXTURE_2D, 0, format, width, height, 0, format, halfFloatOESEnum, data);
     if(!extensionEnabled) {
-        wtu.glErrorShouldBe(gl, gl.INVALID_ENUM, "Half floating point texture must be disallowed if OES_texture_half_float isn't enabled");
+        wtu.glErrorShouldBeIn(gl, [gl.INVALID_ENUM, gl.INVALID_OPERATION],
```
Comment (kenrussell) [inline sdk/tests/conformance/extensions/oes-texture-half-float.html]: INVALID_ENUM is the only error that should be generated here.

File: sdk/tests/conformance/extensions/webgl-depth-texture.html [hunk 5]
```diff
@@ -116,8 +116,10 @@
 
     var tex = gl.createTexture();
     gl.bindTexture(gl.TEXTURE_2D, tex);
-    wtu.shouldGenerateGLError(gl, gl.INVALID_ENUM, 'gl.texImage2D(gl.TEXTURE_2D, 0, gl.DEPTH_COMPONENT, 1, 1, 0, gl.DEPTH_COMPONENT, gl.UNSIGNED_SHORT, null)');
-    wtu.shouldGenerateGLError(gl, gl.INVALID_ENUM, 'gl.texImage2D(gl.TEXTURE_2D, 0, gl.DEPTH_COMPONENT, 1, 1, 0, gl.DEPTH_COMPONENT, gl.UNSIGNED_INT, null)');
+    wtu.shouldGenerateGLErrorIn(gl, [gl.INVALID_ENUM, gl.INVALID_OPERATION],
+                                'gl.texImage2D(gl.TEXTURE_2D, 0, gl.DEPTH_COMPONENT, 1, 1, 0, gl.DEPTH_COMPONENT, gl.UNSIGNED_SHORT, null)');
+    wtu.shouldGenerateGLErrorIn(gl, [gl.INVALID_ENUM, gl.INVALID_OPERATION],
```
Comment (kenrussell) [inline sdk/tests/conformance/extensions/webgl-depth-texture.html]: For these two calls, it looks like INVALID_ENUM and INVALID_VALUE are acceptable errors, but not INVALID_OPERATION.

File: sdk/tests/conformance/extensions/webgl-draw-buffers.html [hunk 6]
```diff
@@ -228,8 +228,8 @@
 
   debug("Testing drawBuffersWEBGL with default drawing buffer");
   shouldBe("gl.getParameter(ext.DRAW_BUFFER0_WEBGL)", "gl.BACK");
-  wtu.shouldGenerateGLError(gl, gl.INVALID_VALUE, "ext.drawBuffersWEBGL([])");
-  wtu.shouldGenerateGLError(gl, gl.INVALID_VALUE, "ext.drawBuffersWEBGL([gl.NONE, gl.NONE])");
+  wtu.shouldGenerateGLErrorIn(gl, [gl.INVALID_VALUE, gl.INVALID_OPERATION], "ext.drawBuffersWEBGL([])");
```
Comment (kenrussell) [inline sdk/tests/conformance/extensions/webgl-draw-buffers.html]: It looks to me like this test has always been buggy, and that it should have produced INVALID_OPERATION, not INVALID_VALUE, since the length of the incoming sequence is 0 (corresponding to the n==0 case in the EXT_draw_buffers spec). Do you agree?

File: sdk/tests/conformance/extensions/webgl-draw-buffers.html [hunk 7]
```diff
@@ -228,8 +228,8 @@
...
+  wtu.shouldGenerateGLErrorIn(gl, [gl.INVALID_VALUE, gl.INVALID_OPERATION], "ext.drawBuffersWEBGL([gl.NONE, gl.NONE])");
```
Comment (kenrussell) [inline sdk/tests/conformance/extensions/webgl-draw-buffers.html]: Similarly, it looks like this one should always have been producing INVALID_OPERATION, not INVALID_VALUE, since n > 1 and the default back buffer is bound. Do you agree?

File: sdk/tests/conformance/textures/misc/texture-formats-test.html [hunk 8]
```diff
@@ -83,9 +83,7 @@
       var err = gl.getError();
       if (err == gl.NO_ERROR) {
           testFailed("should NOT be able to create texture of type " + formatName);
-      } else if (err == gl.INVALID_OPERATION) {
-          testFailed("should return gl.INVALID_ENUM for type " + formatName);
-      } else if (err == gl.INVALID_ENUM) {
+      } else if (err == gl.INVALID_ENUM || err == gl.INVALID_OPERATION) {
```
Comment (kenrussell) [inline sdk/tests/conformance/textures/misc/texture-formats-test.html]: These should only generate INVALID_ENUM or INVALID_VALUE because neither format or internalformat is legal, though they're both the same, per https://www.khronos.org/opengles/sdk/docs/man/xhtml/glTexImage2D.xml .

**Comment by kenrussell** [comment]
@jdashg could you please review the feedback above? Thanks.

File: sdk/tests/conformance/context/context-lost-restored.html (hunk 1 above)
Comment (jdashg) [inline sdk/tests/conformance/context/context-lost-restored.html]: Rereading the specs, I agree. (It actually must be INVALID_OPERATION for WebGL 2)

File: sdk/tests/conformance/extensions/ext-sRGB.html (hunk 2 above)
Comment (jdashg) [inline sdk/tests/conformance/extensions/ext-sRGB.html]: Looks like it for GLES2. GLES3 is ENUM or OPERATION.

File: sdk/tests/conformance/extensions/oes-texture-float.html (hunk 3 above)
Comment (jdashg) [inline sdk/tests/conformance/extensions/oes-texture-float.html]: Rereading the spec for GLES2, I think you are correct.

File: sdk/tests/conformance/extensions/webgl-draw-buffers.html (hunk 7 above)
Comment (jdashg) [inline sdk/tests/conformance/extensions/webgl-draw-buffers.html]: Yeah, I don't see a way for this to be INVALID_VALUE.

File: sdk/tests/conformance/more/functions/texImage2DBadArgs.html [hunk 9]
```diff
@@ -61,7 +61,8 @@
     assertGLError(gl, gl.INVALID_ENUM, "bad target", function(){
         gl.texImage2D(gl.FLOAT, 0, gl.RGBA, 1,1,0,gl.RGBA,gl.UNSIGNED_BYTE, null);
     });
-    assertGLError(gl, gl.INVALID_ENUM, "bad internal format/format", function(){
+    assertGLErrorIn(gl, [gl.INVALID_ENUM, gl.INVALID_OPERATION],
```
Comment (jdashg) [inline sdk/tests/conformance/more/functions/texImage2DBadArgs.html]: I agree.

File: sdk/tests/conformance/more/functions/texSubImage2DBadArgs.html [hunk 10]
```diff
@@ -91,7 +91,7 @@
     assertGLError(gl, gl.INVALID_ENUM, "bad format", function(){
         gl.texSubImage2D(gl.TEXTURE_2D, 0, 0,0, 1,1,gl.FLOAT,gl.UNSIGNED_BYTE, new Uint8Array([0,0,0,0]));
     });
-    assertGLError(gl, gl.INVALID_ENUM, "bad type", function(){
+    assertGLErrorIn(gl, [gl.INVALID_ENUM, gl.INVALID_OPERATION], "bad type", function(){
```
Comment (jdashg) [inline sdk/tests/conformance/more/functions/texSubImage2DBadArgs.html]: I think I agree with this now.

File: sdk/tests/conformance/renderbuffers/framebuffer-test.html [hunk 11]
```diff
@@ -156,16 +156,18 @@
   wtu.glErrorShouldBe(gl, gl.INVALID_ENUM,
             "calling framebufferRenderbuffer with attachment = COLOR_ATTACHMENT1 should generate INVALID_ENUM.");
 
+  gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, fbtex, 0);
+  wtu.glErrorShouldBe(gl, gl.NO_ERROR,
+            "attaching a texture to a framebuffer should succeed.");
+
+  // This must come after the successful attachment, otherwise it could also
+  // generate INVALID_OPERATION.
```
Comment (jdashg) [inline sdk/tests/conformance/renderbuffers/framebuffer-test.html]: Correct, I was reading GetRenderbufferParameteriv.

File: sdk/tests/conformance/extensions/oes-texture-half-float.html (hunk 4 above)
Comment (jdashg) [inline sdk/tests/conformance/extensions/oes-texture-half-float.html]: Yep.

File: sdk/tests/conformance/extensions/webgl-depth-texture.html (hunk 5 above)
Comment (jdashg) [inline sdk/tests/conformance/extensions/webgl-depth-texture.html]: Yep.

File: sdk/tests/conformance/extensions/ext-sRGB.html (hunk 2 above)
Comment (kenrussell) [inline sdk/tests/conformance/extensions/ext-sRGB.html]: Note that this extension is promoted to core and not available in ES3/WebGL 2.0, so the ES3 rules don't apply to this error check.

File: sdk/tests/conformance/extensions/webgl-draw-buffers.html (hunk 7 above)
Comment (kenrussell) [inline sdk/tests/conformance/extensions/webgl-draw-buffers.html]: I've split off this fix into #1318 . The rest will come in another pull request.

File: sdk/tests/js/webgl-test-utils.js [hunk 12]
```diff
@@ -1688,7 +1668,35 @@ var glErrorShouldBe = function(gl, glErrors, opt_msg) {
   glErrorShouldBeImpl(gl, glErrors, true, opt_msg);
 };
 
+/**
+ * Tests that the first error GL returns is in the specified error list.
+ * @param {!WebGLContext} gl The WebGLContext to use.
+ * @param {!Array.<number>} expectedErrorList The list of expected gl errors.
+ * @param {string} opt_msg
+ */
+var glErrorShouldBeIn = function(gl, expectedErrorList, opt_msg) {
```
Comment (kenrussell) [inline sdk/tests/js/webgl-test-utils.js]: Why was this needed? It looks like it has the same basic functionality as glErrorShouldBeImpl, below.
